garage/tf/spaces/dict.py

```diff
@@ -0,0 +1,95 @@
+"""Spaces.Dict for TensorFlow."""
+from collections import Iterable
+from collections import OrderedDict
+import gym
+from gym.spaces import Box as GymBox
+from gym.spaces import Dict as GymDict
+from gym.spaces import Discrete as GymDiscrete
+from gym.spaces import Tuple as GymTuple
+import tensorflow as tf
+
+from garage.misc.overrides import overrides
+from garage.spaces.dict import Dict as GarageDict
+from garage.tf.spaces.box import Box
+from garage.tf.spaces.discrete import Discrete
+from garage.tf.spaces.product import Product
+
+
+class Dict(GarageDict):
+    """TensorFlow extension of garage.Dict."""
+
+    def __init__(self, spaces):
+        assert spaces is not None
+        super().__init__(spaces)
+
+        # Convert each space to a garage.space
+        for key, space in self.spaces.items():
+            self.spaces[key] = self._to_garage_space(space)
+
+    def new_tensor_variable(self, name, extra_dims, flatten=True):
+        return tf.placeholder(
+            dtype=tf.uint8,
+            shape=[None] * extra_dims + [self.flat_dim],
+            name=name)
+
+    @property
+    def dtype(self):
+        """
+        Returns the element's type.
+
+        Returns:
+            dtype (gym.spaces)
+        """
+        return gym.spaces
+
+    @overrides
+    def sample(self):
+        return OrderedDict(
+            [(key, space.sample()) for key, space in self.spaces.items()])
+
+    @property
+    def flat_dim(self):
+        return sum([space.flat_dim for _, space in self.spaces.items()])
+
+    def flat_dim_with_keys(self, keys: Iterable):
+        return sum([self.spaces[key].flat_dim for key in keys])
+
+    def flatten(self, x):
```

Comment:
I think the x is a dict here? Then it should probably be
```
def flatten(self, x):
      return np.concatenate(space.flatten(xs) for space, xs in zip(self.spaces.values(), self.x.values())
```
We need to make sure the space and x has same order.

Reply:
Yes.